#if (!requireNamespace("BiocManager", quietly = TRUE))
#  install.packages("BiocManager")

#BiocManager::install("edgeR")

# Load edgeR module, allow Rstudio to print 100000 lines, and set working directory

library(edgeR)

options(max.print=100000)

setwd("~/Documents/Class_Materials/Fun_Gen/Macaque_Project")

# Read in gene count matrix generated by Stringtie and name species groups

countdata <- read.csv("macaque_gene_count_matrix.csv", row.names="gene_id")

group <- factor(c("Sin","Fas","Arc"))

sum(countdata)
                
# Run edgeR functions including generating a matrix of differential expression patterns for genes based on species comparisons
# Filter out genes with low expression and apply normalization for sequencing library sizes

y <- DGEList(counts=countdata,group=group)

keep <- filterByExpr(y)

summary(y)

y <- y[keep,,keep.lib.sizes=FALSE]

summary(y)

y <- calcNormFactors(y)

summary(y)

design <- model.matrix(~group)

# Conduct exact tests of expression differences using arbitrary dispersion value and print results to text files

SinArc <- exactTest(y, pair=c("Sin","Arc"), dispersion=0.16)
FasArc <- exactTest(y, pair=c("Fas","Arc"), dispersion=0.16)
SinFas <- exactTest(y, pair=c("Sin","Fas"), dispersion=0.16)

sink("SinArcDiffExpr.txt")
SinArc
sink()

sink("FasArcDiffExpr.txt")
FasArc
sink()

sink("SinFasDiffExpr.txt")
SinFas
sink()

# Read in genital genes extracted from the above files
# This was done on a command line using a simple grep command on a candidate gene list

GeniGenes_SinArc_data=read.table("GeniGenesSinArc.txt")
GeniGenes_FasArc_data=read.table("GeniGenesFasArc.txt")
GeniGenes_SinFas_data=read.table("GeniGenesSinFas.txt")

# Read in random replicated genes
# These were generated on a command-line using the GNU shuf tool with a command like:
# for x in {1..100}; do shuf -n 77 SinArcDiffExpr.txt > SinArc_${x}.txt; done

SinArc_data=read.table("Combined_SinArc.txt")
FasArc_data=read.table("Combined_FasArc.txt")
SinFas_data=read.table("Combined_SinFas.txt")

# Generate histograms for the above data sets

hist(SinArc_data$V2, col = '#a6dba0', breaks = 42, 
     main="Bear compared to Tibetan macaque", xlab = "LogFC Expression")
abline(v = mean(SinArc_data$V2),                       
       col = "red",
       lwd = 3)
abline(v = mean(GeniGenes_SinArc_data$V2),                       
       col = "blue",
       lwd = 3)

hist(FasArc_data$V2, col = '#fdb863', breaks = 42,
     main="Bear compared to rhesus macaque", xlab = "LogFC Expression")
abline(v = mean(FasArc_data$V2),                       
       col = "red",
       lwd = 3)
abline(v = mean(GeniGenes_FasArc_data$V2),                      
       col = "blue",
       lwd = 3)

hist(SinFas_data$V2, col = '#c2a5cf', breaks = 42,
     main="Rhesus compared to Tibetan macaque", xlab = "LogFC Expression")
abline(v = mean(SinFas_data$V2),                       
       col = "red",
       lwd = 3)
abline(v = mean(GeniGenes_SinFas_data$V2),                       
       col = "blue",
       lwd = 3)

# Conduct t-tests comparing genital genes to replicated genes

t.test(GeniGenes_SinArc_data$V2,SinArc_data$V2)
t.test(GeniGenes_FasArc_data$V2,FasArc_data$V2)
t.test(GeniGenes_SinFas_data$V2,SinFas_data$V2)

# Use decideTests with default FDR=0.05 to isolate genes with significant expression differences

FDR_SinArc=decideTests(SinArc)
FDR_FasArc=decideTests(FasArc)
FDR_SinFas=decideTests(SinFas)

write.table(FDR_SinArc, file="Sig_Expression_SinArc.txt")

write.table(FDR_FasArc, file="Sig_Expression_FasArc.txt")

write.table(FDR_SinFas, file="Sig_Expression_SinFas.txt")

# Genital genes were later extracted from these lists on a command line
